lib/stdlib/src/binary.erl

```diff
@@ -934,6 +935,34 @@ unhex(X) ->
              no, 10, 11, 12, 13, 14, 15, no, no, no, no, no, no, no, no, no  %96
             }).
 
+-doc """
+Joins a list of binaries together by a specified `Separator`.
+
+Equivalent to `iolist_to_binary(lists:join(Separator, Binaries))` with `binary:join/2` being the faster of the two.
+
+_Example:_
+
+```erlang
+1> binary:join([<<"a">>, <<"b">>, <<"c">>], <<", ">>).
+<<"a, b, c">>
+```
+""".
+-doc(#{since => <<"OTP 27.0">>}).
+-spec join([binary()], binary()) -> binary().
+join([], _Separator) -> <<>>;
+join([H], Separator) when is_binary(H) andalso is_binary(Separator) -> H;
+join([H | T], Separator) when is_binary(Separator) ->
+    %% Starting with an empty binary convinces the compiler to use the new "private append" optimisation
+    Acc = <<>>,
+    join(T, Separator, <<Acc/binary, H/binary>>);
```

Comment:
@bjorng @jhogberg @michalmuskala 

I understand why starting with the empty string makes the compiler use the new "private append" optimisation. However I wonder if this could be generalized?

While `H` cannot be privately appended because it comes from an external function, the next iteration of `join/3` can be privately appended, because no one uses the intermediate binary. Therefore, for those "closed loops", would it make sense to have a bit in the binary that tells when to private append or not? Generally speaking, everything after the first iteration would be privately appended. This would allow private append to happen in more situations, although I am not aware of the costs of reserving one extra bit for binaries.

PS: I may be completely off mark here.

Reply:
> I understand why starting with the empty string makes the compiler use the new "private append" optimisation. However I wonder if this could be generalized?
> 
> While `H` cannot be privately appended because it comes from an external function, the next iteration of `join/3` can be privately appended, because no one uses the intermediate binary.

The private append operation is faster because it does fewer tests and less working than general append operation. Therefore, the private append **must** only be used with a binary that has been specially prepared.

However, it should be possible for the compiler to generalize the optimization. If the call looks like:

```
join(T, Separator, H)
```

the compiler could rewrite it to:

```
    Acc = <<>>,
    join(T, Separator, <<Acc/binary, H/binary>>);
```

Not sure that this code pattern is common enough to make the optimization worthwhile to implement, though. It would not be needed if the clause is rewritten to:

```
join([_ | _]=List, Separator) when is_binary(Separator) ->
    join(List, Separator, <<>>);
```